### PR TITLE
feat: add SSR HTTP Client Hints support

### PR DESCRIPTION
**Pull request by userquin**
This PR will be merged, includes: ~~This PR will not be merged here (will go to Nuxt or Nitro), it is a small test for [Http Client Hints](https://developer.mozilla.org/en-US/docs/Web/HTTP/Client_hints#user_preference_media_features_client_hints) for SSR (will work only on Chromium based browsers, tested only in Chrome and Edge on Windows)~~:
- [Critical-CH](https://wicg.github.io/client-hints-infrastructure/#critical-ch-algo)
- [Sec-CH-Prefers-Color-Scheme](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-CH-Prefers-Color-Scheme)
- [Sec-CH-Prefers-Reduced-Motion](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-CH-Prefers-Reduced-Motion)
- [Sec-CH-Viewport-Width](https://wicg.github.io/responsive-image-client-hints/#sec-ch-viewport-width)
- [Sec-CH-Viewport-Height](https://wicg.github.io/responsive-image-client-hints/#sec-ch-viewport-height)

Check also [Client Hints Infrastructure](https://wicg.github.io/client-hints-infrastructure/).

/cc @danielroe

### CODE REVIEW

File: playground/server/plugins/accept-ch.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import type { NitroAppPlugin, RenderResponse } from 'nitropack'
+import { addClientHintResponseHeaders } from '~/utils/client-hints'
+
+export default <NitroAppPlugin> function (nitro) {
```
Comment (userquin) [inline playground/server/plugins/accept-ch.ts:4]: Move this to the middleware, add user agent detection to include headers only for chromium based browsers, add vary on response and include ibfo in route meta.

**Comment by userquin** [comment]
@mubaidr I've the viewport with and height working properly on my local, since I'm adding also `prefers-color-scheme` but not configuring it, we can give it a try later once I pushed my latest changes, maybe we won't need your hack anymore...

I guess we need to add a cookie, since we only have it on initial request, if using some switcher we'll need to restore the value from the cookie and don't sent the hint header once configured, ofc, we should change the cookie. Maybe we can use some state instead the cookie, I'm using it for the values...

**Comment by userquin** [comment]
@danielroe  @mubair full viewport support added, I'll check adding auto theme support, only added support for headers, I guess we need to include a cookie for it

**Comment by userquin** [comment]
@mubaidr wrong name in previous comment...

File: src/runtime/plugins/vuetify-ssr-client-hints.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  const event = useRequestEvent(nuxtApp)
+
+  if (!event) {
+    if (clientHints.viewportSize) {
+      // on client, we update the display and watch for changes to avoid hydration mismatch on page refresh
+      // there will be some hydration mismatch since the headers sent by the user agent may not be accurate
+      nuxtApp.hook('app:beforeMount', () => {
+        const { width, height, update } = useNuxtApp().$vuetify.display
+        update()
+        watch(() => [width.value, height.value], ([width, height]) => {
```
Comment (userquin) [inline src/runtime/plugins/vuetify-ssr-client-hints.ts:18]: is this required? it seems if we don't update the screem (in the playground) we get hydration mismatch error, I'll test it again.

Comment (userquin) [reply src/runtime/plugins/vuetify-ssr-client-hints.ts:18]: maybe we only need the `update` call

**Comment by userquin** [comment]
@danielroe @jd-solanki @mubraid added two theme support, include simple example in the playground, it is working! Also update the docs.

I need to include browser detection and add support for browsers not supporting the HTTP Client Hints, we'll get a lot of mismatch hydration warnings on page reload

**Comment by userquin** [comment]
@mubaidr wrong name in previous comment...

**Comment by userquin** [comment]
The PR has been updated with the following:
- browser detection
- fix viewport size mismatch hydration warning in browser that doesn't support client hints: will use the ssr configuration and once hydrated the `useDisplay` will update the ui; on browsers supporting client hints there is no need, the state will be restored properly

I need to include some new browsers like Samsung Browser, Arc and Brave, and check iOS/macOS Chromium based browsers, maybe we need to exclude them. I cannot test since I haven't my macOS laptop, so any help will be appreciate.

**Comment by userquin** [comment]
`document.cookie` is a pain, on FF will not work or will show a few warnings, maybe we need to include an api call
on page refresh the server will set the cookie in Firefox.

imagen

If I toogle the theme, we change document.cookie, check the result in FF:

imagen

On chromium based browsers works as expected. I mean, it only change the value and not the attributes, whereas in FF, it will change also the attributes.

**Comment by mubaidr** [comment]
Just curious, when using simple composable, there is no issue with hydration in my app with SSR. Why do we need client hints? Am I missing something?

**Comment by userquin** [comment]
> Just curious, when using simple composable, there is no issue with hydration in my app with SSR. Why do we need client hints? Am I missing something?

Your composable is just a workaround to avoid mismatch hydration warning/error, you can use `useState` in your composable to achieve also the proper behaviour, you're using `localStorage` and the server will not have access to it.

Anyway, this PR is about having the same data in the client and the server. Adding this is about receiving the page content from the SSR response with the theme applied, that's, the page will be rendered properly (dark/light or any other theme applied).

**Comment by mubaidr** [comment]
Doh! Totally forgot about that issue. Awesome work on that one, looking forward.

**Comment by userquin** [comment]
> Doh! Totally forgot about that issue. Awesome work on that one, looking forward.

IIRC, your composable was the workaround to restore the selected theme properly. If you don't restore the theme, then there is no hydration mismatch at all, you're refreshing the page.

**Comment by userquin** [comment]
I'm splitting the client hints detection, every hint will be available depending on the browser, OS and version.

**Comment by userquin** [comment]
@mubaidr @jd-solanki PR updated including:
- split client hint detection using the browser, OS and version form the user agent header
- add reload page in first request support: by default it is disabled: Nuxt client plugin provides the logic to reload the page: the Nuxt server plugin will omit the theme cookie if it is the first request when required (check the `shouldWriteCookieOnInitialRequest  function)
- the playground now has the proper dark theme, removed light v-sheet color in the layout
- docs updated properly: included the new reload page and added warning about browsers supporting the client hints

**Pull request merged**